### ISSUE TITLE
The GPU softCrossEntropy does not validate teacher ids, where the CPU one does

### ISSUE THREAD

**Issue by felladrin** [felladrin/gguf-trainer#61]
Found while fixing #55, which covers `crossEntropy` and `fusedCrossEntropy` on both backends.

The CPU `softCrossEntropy` validates each teacher id before using it (`src/model/autograd.ts`):

```ts
if (id < 0 || id >= V) throw new Error(`softCrossEntropy: teacher id ${id} out of [0,${V})`);
```

The GPU one does not (`src/backend/webgpu.ts`, the `softCrossEntropy` method). It uploads `teacherIds` straight to the kernel and counts kept rows without looking at the ids, so the same input that stops on the CPU runs to completion on the device.

That is #55's inconsistency mirrored: there, the loss with the check was `softCrossEntropy` and the one without it was on every training path; here it is the same op differing by backend, which is the divergence the parity suite exists to prevent. The practical impact is smaller, since `softCrossEntropy` is the Phase B KL anchor and its ids come from a precomputed teacher file rather than from a corpus, but a teacher file built against a different vocab is the same class of mistake.

`keptRowsInVocab` (added in #55) does not fit as-is: `teacherIds` is `[T, k]` and the row gate is `teacherIds[t * k] < 0`, not per-entry. Either a small variant of it, or the same inline check the CPU already has, placed in the GPU method's existing kept-count loop so it costs no extra pass.

Whatever lands wants a `gpu-parity.ts` arm in the shape of `targetRangeGate`: the CPU-only check means a device arm is the only thing that would catch a revert.

**Comment by felladrin** [comment]
A structural way to close this, raised while reviewing #69.

The reason this gap exists is that `crossEntropy` and `softCrossEntropy` validate *below* the backend dispatch, so every implementation needs its own call and the GPU one was the implementation nobody remembered. `embedding` (#63) validates above the dispatch instead, which makes that class of omission unreachable: there is one call site and both backends inherit it.

So the fix here could be the hoist rather than a fourth call. Move the teacher-id check in `softCrossEntropy` above `if (opsBackend) return opsBackend.softCrossEntropy(...)`, and the GPU path is covered by construction.

The cost is the same one it has everywhere: the CPU version folds its check into a loop it already runs, so hoisting makes it a separate O(T*k) pass, and the GPU version would then either duplicate the kept-row count or take it as an argument. Worth weighing against the fourth-call version rather than assuming the hoist wins.
